pkg/operator/webhookdeployment/webhook.go

```diff
@@ -0,0 +1,162 @@
+		eventRecorder:           eventRecorder,
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-snapshot-controller"),
+		operandVersion:          operandVersion,
+		csiSnapshotWebhookImage: csiSnapshotWebhookImage,
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(client).WithInformers(
+		client.Informer(),
+		deployInformer.Informer(),
+	).ToController(WebhookControllerName, eventRecorder)
+}
+
+func (c *csiSnapshotWebhookController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, opStatus, _, err := c.client.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	deployment, err := c.getDeployment(opSpec)
+	if err != nil {
+		// This will set Degraded condition
+		return err
+	}
+	lastGeneration := resourcemerge.ExpectedDeploymentGeneration(deployment, opStatus.Generations)
+	deployment, _, err = resourceapply.ApplyDeployment(c.kubeClient.AppsV1(), c.eventRecorder, deployment, lastGeneration)
+	if err != nil {
+		// This will set Degraded condition
+		return err
+	}
+
+	// Compute status
+	// Available: at least one replica is running
+	deploymentAvailable := operatorapi.OperatorCondition{
+		Type: WebhookControllerName + operatorapi.OperatorStatusTypeAvailable,
+	}
+	if deployment.Status.AvailableReplicas > 0 {
+		deploymentAvailable.Status = operatorapi.ConditionTrue
+	} else {
+		deploymentAvailable.Status = operatorapi.ConditionFalse
+		deploymentAvailable.Reason = "WaitDeployment"
+		deploymentAvailable.Message = "Waiting for a Deployment pod to start"
+	}
+
+	// Not progressing: all replicas are at the latest version && Deployment generation matches
+	deploymentProgressing := operatorapi.OperatorCondition{
+		Type: WebhookControllerName + operatorapi.OperatorStatusTypeProgressing,
+	}
+	if deployment.Status.ObservedGeneration != deployment.Generation {
+		deploymentProgressing.Status = operatorapi.ConditionTrue
+		deploymentProgressing.Reason = "NewGeneration"
+		msg := fmt.Sprintf("desired generation %d, current generation %d", deployment.Generation, deployment.Status.ObservedGeneration)
+		deploymentProgressing.Message = msg
+	} else {
+		if deployment.Spec.Replicas != nil {
+			if deployment.Status.UpdatedReplicas == *deployment.Spec.Replicas {
+				deploymentProgressing.Status = operatorapi.ConditionFalse
+				// All replicas were updated, set the version
+				c.versionGetter.SetVersion(webhookVersionName, c.operandVersion)
+			} else {
+				msg := fmt.Sprintf("%d out of %d pods running", deployment.Status.UpdatedReplicas, *deployment.Spec.Replicas)
+				deploymentProgressing.Status = operatorapi.ConditionTrue
+				deploymentProgressing.Reason = "WaitDeployment"
+				deploymentProgressing.Message = msg
+			}
+		}
+	}
+
+	updateGenerationFn := func(newStatus *operatorapi.OperatorStatus) error {
+		if deployment != nil {
```

Comment:
Would it have panicked above if `deployment` were nil?

Reply:
It would :-)
Removed the check... What could possibly go wrong, right? :-D